simpleclient/src/main/java/io/prometheus/client/CKMSQuantiles.java

```diff
@@ -217,46 +306,83 @@ private boolean insertBatch() {
+     * <p>
+     * Compress. Periodically, the algorithm scans the data
+     * structure and merges adjacent nodes when this does not
+     * violate the invariant. That is, remove nodes (vi, gi, ∆i)
+     * and (vi+1 , gi+1 , ∆i+1 ), and replace with (vi+1 , (gi +
+     * gi+1 ), ∆i+1 ) provided that (gi + gi+1 + ∆i+1 ) ≤ f (ri , n).
+     * This also maintains the semantics of g and ∆ being the
+     * difference in rank between v_i and v_{i-1} , and the difference
+     * between the highest and lowest possible ranks of vi, respectively.
      */
     private void compress() {
-        if (sample.size() < 2) {
+        // If there are 0,1, or 2 samples then there's nothing to compress.
+        if (sample.size() < 3) {
             return;
         }
 
         ListIterator<Item> it = sample.listIterator();
-        int removed = 0;
 
-        Item prev = null;
+        Item prev;
         Item next = it.next();
 
+        // Counter for the rank in the stream of all observed values.
+        // Paper Section 3.1 on true rank: let r_i = Sum_{j=1}^{i−1} g_j
+        int currentRank = next.g;
+
         while (it.hasNext()) {
             prev = next;
             next = it.next();
-
-            if (prev.g + next.g + next.delta <= allowableError(it.previousIndex())) {
+            currentRank += next.g;
+            if (prev.g + next.g + next.delta <= allowableError(currentRank)) {
                 next.g += prev.g;
                 // Remove prev. it.remove() kills the last thing returned.
-                it.previous();
-                it.previous();
-                it.remove();
-                // it.next() is now equal to next, skip it back forward again
-                it.next();
-                removed++;
+                it.previous(); // brings pointer back to 'next'
+                it.previous(); // brings pointer back to 'prev'
+                it.remove(); // remove prev
+                // it.next() is now equal to next, 
+                it.next(); // set pointer to 'next'
             }
         }
     }
 
-    private class Item {
+    /**
+     * As in GK, the data structure at time n, S(n), consists
+     * of a sequence of s tuples ⟨ti = (vi, gi, ∆i)⟩, where each vi
+     * is a sampled item from the data stream and two additional
+     * values are kept: (1) g_i is the difference between the lowest
+     * possible rank of item i and the lowest possible rank of item
+     * i − 1; and (2) ∆_i is the difference between the greatest
+     * possible rank of item i and the lowest possible rank of item
+     * i.
+     */
+    private static class Item {
+        /**
+         * vi
+         * is a sampled item from the data stream and two additional
+         * values are kept
+         */
         public final double value;
+        /**
+         * g_i is the difference between the lowest
+         * possible rank of item i and the lowest possible rank of item
+         * i − 1
+         */
         public int g;
```

Comment:
Again not your code, but `g` should be `final` as well.

Reply:
This `g` is written to in the compress method. `next.g += prev.g;`. I will remove the public from the members in `Quantile` and `Item` though.